### PR TITLE
feat(frontend): add extension modifier parsing

### PR DESCRIPTION
**Pull request by SOF3**
### Description

Frontend merges traces from additional sources

image

### CODE REVIEW

File: pkg/util/semaphore/semaphore.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+	var stopped atomic.Bool
+	defer stopped.Store(true)
+
+	for {
+		select {
+		case <-doneCh:
+			// publishCh must be fully sent before calling sem.doneWg.Done()
+			return nil
+		case publish := <-sem.publishCh:
+			err := publish()
```
Comment (xuqingyun) [inline pkg/util/semaphore/semaphore.go:144]: why publishes results after calling Run() function, rather than just run it after task finish immediately

**Review by xuqingyun — APPROVED** [review]
lgtm

File: pkg/util/semaphore/semaphore.go (hunk 1 above)
Comment (xuqingyun) [inline pkg/util/semaphore/semaphore.go:144]: why publishes results after calling Run() function, rather than just run it after task finish immediately

Comment (SOF3) [reply pkg/util/semaphore/semaphore.go:144]: Thread safety. `publish()` is always called from the main goroutine that calls `Run()`, so we don't have to acquire a separate lock for publishing the results (which is typically appending data to a slice/map).

**Pull request merged**